### PR TITLE
Example/pi/can sender

### PR DESCRIPTION
**Pull request by Burke-Daniel**
This pull request contains two python programs, one for sending can messages, and one for receiving. It also contains json configurations for both programs, so that the user can configure the messages, filters, and bitrate to their preferences/needs. 

(Putting this PR up prematurely so you can have a look at the meat of the programs while I make some changes/ask for your opinion)

### CODE REVIEW

File: Pidaq/Examples/can-sender.py
Comment (Burke-Daniel) [inline Pidaq/Examples/can-sender.py]: @colton-smith is there a better way to keep the program running than this? `send_periodic` will run until the program goes out of scope

File: Pidaq/Examples/receiverInfo.json
Comment (Burke-Daniel) [inline Pidaq/Examples/receiverInfo.json:6]: The python-can library uses hex values for can ID and can mask, but since python represents hex as a string, you pass the decimal representation of the hex value, and it does the conversion

File: Pidaq/Examples/senderInfo.json
Comment (Burke-Daniel) [inline Pidaq/Examples/senderInfo.json:7]: See my comment on receiverInfo.json for how python-can interprets the integer values

**Comment by Burke-Daniel** [comment]
[Create CAN sender program](https://trello.com/c/j5ljeuuD/3-create-can-sender-program)

File: Pidaq/Examples/can-sender.py
Comment (colton-smith) [reply Pidaq/Examples/can-sender.py]: So `send_periodic` runs in the background until the program goes out of scope?  If so I don't think there is a better way to just hang a program. Is there a way to just send a single message and then loop that? 

Edit: would also be good to comment and describe why we are looping endlessly

File: Pidaq/Examples/can-receiver.py
Comment (colton-smith) [inline Pidaq/Examples/can-receiver.py]: Don't need an empty line between the """ and the last line of the doc string

File: Pidaq/Examples/can-receiver.py
Comment (colton-smith) [inline Pidaq/Examples/can-receiver.py]: Remove empty line

File: Pidaq/Examples/can-sender.py
Comment (colton-smith) [inline Pidaq/Examples/can-sender.py]: As per pep-8 and my supervisor at the river - technically there should be an empty line between these two imports as CAN is a third party library and json is a first party library

File: Pidaq/Examples/can-sender.py
Comment (colton-smith) [reply Pidaq/Examples/can-sender.py]: Generally,  the "with" context manager is used when opening files, if you don't use it, the file doesn't get closed until after the file object is garbage collected (which can lead to some undesired behavior). In this case, it wouldn't have any negative effects though- just something to think about.

File: Pidaq/Examples/can-receiver.py
Comment (colton-smith) [inline Pidaq/Examples/can-receiver.py:3]: See below comment about imports

File: Pidaq/Examples/can-receiver.py
Comment (colton-smith) [inline Pidaq/Examples/can-receiver.py:41]: Maybe add a comment to explain how the can_bus object has messages? Does it read the bus periodically and update a list of messages? 

Also, why not have the receiver persistently read the bus in a while loop? From the looks of it, it just reads whatever messages or currently on the bus.

File: Pidaq/Examples/can-sender.py
Comment (colton-smith) [inline Pidaq/Examples/can-sender.py]: Remove this line

File: Pidaq/Examples/can-sender.py
Comment (colton-smith) [inline Pidaq/Examples/can-sender.py]: PEP8 - remove this line

**Review by colton-smith — CHANGES_REQUESTED** [review]
Just some basic style changes, and maybe add a few clarifying comments for things that aren't intuitive.

The style looks real nice apart from those few things

File: Pidaq/Examples/can-sender.py
Comment (Burke-Daniel) [reply Pidaq/Examples/can-sender.py]: `send_periodic` just runs until it goes out of scope yeah, but if I use the `send` to send a single message, then I guess I would have to keep track of time, and then if messages had different frequencies then that might get a bit weird to implement

File: Pidaq/Examples/can-receiver.py
Comment (Burke-Daniel) [reply Pidaq/Examples/can-receiver.py:41]: There are two ways to read the bus for messages, one way is to use the `recv()` method, and just have an infinite loop continuously reading the bus, or the other way is the way that I have done, which is functionally the same.

**Comment by Burke-Daniel** [comment]
@colton-smith I'll get you to take another look now that I think I've addressed your comments

File: Pidaq/Examples/can-receiver.py
Comment (colton-smith) [inline Pidaq/Examples/can-receiver.py]: Put a space between these 2 lines and you're GTM

**Review by colton-smith — COMMENTED** [review]
One last whitespace fix

**Comment by Burke-Daniel** [comment]
Everything should be good now, I'll get you to take one last look at the new commits and the README, then I'll let the boys have a look before I merge

**Comment by Burke-Daniel** [comment]
@khpandya @Ahmed-Hassan21 this is the pull request I was talking about on the call over the weekend, take a look through the files and commits, make comments on anything you think needs to be changed/stuff you are unsure of, and then leave a review comment either approving the PR or requesting changes!

**Review by colton-smith — APPROVED** [review]
GTM

File: Pidaq/Examples/can-receiver.py
Comment (khpandya) [inline Pidaq/Examples/can-receiver.py:34]: not sure if this is proper styling but starting a list bracket on a line, writing code on the next and closing it in the second line looks weird-shouldn't it either be both brackets on the same line or opening bracket, code on the next line and then the closing bracket on another new line?

Comment (khpandya) [inline Pidaq/Examples/can-receiver.py:34]: not sure if this is proper styling but starting a list bracket on a line, writing code on the next and closing it in the second line looks weird-shouldn't it either be both brackets on the same line or opening bracket, code on the next line and then the closing bracket on another new line?

Comment (khpandya) [reply Pidaq/Examples/can-receiver.py:34]: Also, can you please tell me how you select the lines you're reviewing-seems a bit confusing and I have 31-34 highlighted even though I wanted to highlight 34 and 35

Comment (colton-smith) [reply Pidaq/Examples/can-receiver.py:34]: @khpandya Good observation, I believe Daniel added a line break for 2 purposes, adherence to 80 char lines and also readability of the list comprehension. As far as the formatting of the line break goes,  PEP-8 addresses this under "Code Layout" here: <https://www.python.org/dev/peps/pep-0008/#tabs-or-spaces>

" When using a hanging indent the following should be considered; there should be no arguments on the first line and further indentation should be used to clearly distinguish itself as a continuation line: "

Comment (Burke-Daniel) [reply Pidaq/Examples/can-receiver.py:34]: @khpandya Colton is absolutely right, this is just a formatting thing, and is good practice for longer lists/objects/dicts. Good of you to notice. Also regarding your comment about the lines you are reviewing, it is just showing a couple of lines above for context, that is the proper way to select lines.

**Pull request merged**
